add ROR as an Author Identifier Type

pdurbin opened the pull request:
**What this PR does / why we need it**:

This PR adds ROR as an Author Identifier Type.

**Which issue(s) this PR closes**:

- Closes #11075
- Closes #11149

**Special notes for your reviewer**:

As of 1fe95d9 a second regex for ROR was added. Now we have "ROR" to match just the identifier and "ROR_FULL_URL" to match the full ROR URL. The latter is used when exporting the "Datacite" format.

I have a question. How can a ROR uniquely identify an author? If I use the ROR for Harvard ( https://ror.org/03vek6s52 ), how can someone know which of the thousands of Harvard affiliates I am? 🤔 

![Screenshot 2024-12-19 at 3 32 22 PM](https://github.com/user-attachments/assets/b3516fc6-1c1e-4709-b7f6-a0b98a8da1ef)

~~Also, is it ok that ROR has been added to the bottom of the list? That's what we've always done before. Here's how it looks:~~

![Screenshot 2024-12-19 at 3 31 46 PM](https://github.com/user-attachments/assets/dd216021-1c1e-40b9-9fbf-df95b7a84b09)

Nevermind. This was fixed in ddb6bc9. ROR is now after ORCID:

![Screenshot 2024-12-20 at 11 50 35 AM](https://github.com/user-attachments/assets/95a4f637-76e0-45a8-9ff7-5239420ea0f8)


**Suggestions on how to test this**:

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Yes, see screenshots above.

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

Preview at https://dataverse-guide--11118.org.readthedocs.build/en/11118/user/dataset-management.html#adding-a-new-dataset

qqmyers commented:
The intent is to allow Organizations to be authors. As you say, it would not make sense to use a ROR for a person. Eventually an external vocab script would start by letting you select Person or Org, but until then, perhaps @jggautier might want changes to the title  or help tip to make things clearer?

pdurbin commented:
@qqmyers yes, good idea. I do think the tooltip should be edited to account for organizations as authors.

@jggautier do you want me to take a whack at that? Or do you want to suggest something?

jggautier commented:
Hey @pdurbin

> I have a question. How can a ROR uniquely identify an author? If I use the ROR for Harvard ( https://ror.org/03vek6s52 ), how can someone know which of the thousands of Harvard affiliates I am? 🤔

Hmmm, a ROR can uniquely identifier an author when that author is an organization, such as Harvard University.

When you ask "how can someone know which of the thousands of Harvard affiliates I am?", this makes me think that you think that the Identifier Type and Identifier fields are supposed to be used for the identifier of the author's organizational affiliation, in your example Harvard University. Is that right?

It's telling that you think that, because others might, too. During testing after implementation I'll make sure we watch out for this.

We're expecting that if a depositor selects ROR as the Identifier Type, that means that the author is Harvard University, not that the author's affiliation is Harvard University.

> Also, is it ok that ROR has been added to the bottom of the list? That's what we've always done before.
Could you write a bit more about we've always done that? I imagine there are some cases where we think it'll be better to list options in the list alphabetically or by how often we see people selecting them. Or in this case, since we want to encourage the use of ROR over other types of identifiers, i the mockup at https://github.com/IQSS/dataverse/issues/11075 I added it below ORCID.

pdurbin commented:
@jggautier I think we wrote at the same time. Do you want me to try editing the tooltip? Right now it only applies to people, not organizations:

![Screenshot 2024-12-19 at 3 32 22 PM](https://github.com/user-attachments/assets/4b43db99-1fea-4eb7-bda1-81e3fff1d9a7)

jggautier commented:
Ah, I'm not sure how the Identifier Type field's tooltip applies to people and not organizations. Could you write why you think that? Or are you referring to another tooltip?

pdurbin commented:
I'm saying we could change...

"The type of identifier that uniquely identifies the author (e.g. ORCID, ISNI)"

...to something like this:

"The type of identifier that uniquely identifies the author as a person (e.g. ORCID, ISNI) or organization (e.g. ROR)"

jggautier commented:
Ah I see. So you think that we should note that the author can be a person or an organization because other folks might not realize that they can add an organization? I think we should keep an eye on this as we get more feedback and that we shouldn't change the tooltip in this PR.

And kind of related, in [our metadata text guidelines](https://docs.google.com/document/d/1tY5t3gjrIgAGoRxVMWQSCh46fnbSmnFDLQ7aLkNLhJ8) we say not to "include examples in the tooltip text for dropdown fields, since the dropdown fields already include all options". That's the 13th thing on the list. I'm surprised now that this wasn't done when we applied the guidelines to most of the fields in the citation metadata block a few years back as part of https://github.com/IQSS/dataverse/issues/8127.

pdurbin commented:
I'd say I go in thinking "Author... what is an author?... it's a person... like the author of a book... or an author of an article. I see that my name is auto-populated as the author."

"Affiliation makes sense. What's next?" I think.

I get down to "Identifier Type" and see "The type of identifier that uniquely identifies the author (e.g. ORCID, ISNI)".

I think, "ORCID, yes, this will uniquely identify an author." (Again, I'm thinking of an author as a person.)

Then I see "ROR" in the list and I think "a ROR cannot uniquely identify a person. Why is ROR in this list of identifier types? Is this a bug?"

The "Author", "Name", and "Affiliation" tooltips all help reduce confusion by stating that an author can be an organization but the "Identifier Type" tooltip does not. So, I'm simply suggesting we fix this, not necessarily in this pull request, but at some point.

pdurbin commented:
> > Also, is it ok that ROR has been added to the bottom of the list? That's what we've always done before.
> 
> Could you write a bit more about we've always done that? I imagine there are some cases where we think it'll be better to list options in the list alphabetically or by how often we see people selecting them. Or in this case, since we want to encourage the use of ROR over other types of identifiers used for organizations, in the mockup at #11075 I added it below ORCID.

Yes, I noticed that. I wasn't sure if it was safe to simply change the displayOrder or not. This is what I had in this PR originally, with ROR at the bottom (displayOrder=8):

![Screenshot 2024-12-20 at 11 11 18 AM](https://github.com/user-attachments/assets/b11c514a-526f-4239-8d25-b37aee37b546)

However, I just went through an upgrade scenario and it looks like reloading a metadata block caused new displayOrder values to be saved without having an ill effect on older data. (I think I may have known this once but wasn't sure. Perhaps we should document it.)

Here's how I tested it. First, while running the "develop" branch, I added a couple authors with ORCID (first) and ISNI (second):

![Screenshot 2024-12-20 at 11 48 34 AM](https://github.com/user-attachments/assets/acfe8b5c-676c-41e8-8192-38dc86c63950)

Then I edited citation.tsv to put ROR after ORCID as shown in the mockup. Since it worked (see below) I went ahead and pushed this change: 029c247

Then I reloaded citation.tsv:

`curl http://localhost:8080/api/admin/datasetfield/load -H "Content-type: text/tab-separated-values" -X POST --upload-file citation.tsv`

Now ROR appears second, right after ORCID:

![Screenshot 2024-12-20 at 11 50 35 AM](https://github.com/user-attachments/assets/ea6e8892-ce3d-463d-ac4f-c37b4e747fac)

I checked the database and (as expected) ROR has a higher ID than other Author Identifiers (since it was added later) but the sorting looks right, looks like what we want:

```
dataverse=# select * from controlledvocabularyvalue where datasetfieldtype_id = 11 order by displayorder;
  id  | displayorder | identifier |   strvalue   | datasetfieldtype_id 
------+--------------+------------+--------------+---------------------
   52 |            0 |            | ORCID        |                  11
 8569 |            1 |            | ROR          |                  11
   53 |            2 |            | ISNI         |                  11
   54 |            3 |            | LCNA         |                  11
   55 |            4 |            | VIAF         |                  11
   56 |            5 |            | GND          |                  11
   57 |            6 |            | DAI          |                  11
   58 |            7 |            | ResearcherID |                  11
   59 |            8 |            | ScopusID     |                  11
(9 rows)
```

So, I think we're good. We'll see what people say in review and QA. 😅 

Again, I'm happy to document the behavior or reloading the metadata blocks to say you can reorder controlled vocabulary values. I was happily surprised. 😄

jggautier commented:
Writing in the guides that we're able to reorder controlled vocabulary values sounds good to me. I've always assumed it was possible just by changing the displayOrder number, but saying so explicitly could other others who assumed it wasn't possible. Let me know if I can help!

pdurbin commented:
@jggautier at standup this morning @qqmyers asked if you're ready for this to move forward and I said you are. If that's not right please pull this PR out of Ready for Review. Thanks!

jggautier commented:
Thanks @pdurbin. I don't see changes to any files related to the Dataverse Guides.

Have you decided not to add to the guides that we're able to reorder controlled vocabulary values?

If so, yeah I'm fine with this moving forward.

jggautier commented:
Ah yeah. I don't feel strongly either way. And thinking about it more I don't think it's worth trying to learn more. I think the note you added makes sense.

qqmyers reviewed (approved):
Looks fine

ofahimIQSS commented:
a separate issue has been created related to the incorrect handling of ROR identifiers in the dataset metadata: https://github.com/IQSS/dataverse/issues/11149

ofahimIQSS commented:
Description
While testing the pull request [#11118](https://github.com/IQSS/dataverse/pull/11118), an issue was identified with the handling of ROR (Research Organization Registry) identifiers in the Dataverse system. Specifically, the system generates incorrect links when saving and displaying ROR identifiers in the dataset metadata.

**Steps to Reproduce**
Follow the steps outlined in [PR #11118](https://github.com/IQSS/dataverse/pull/11118).
Create a new dataset and proceed to the "Author" section:
Under Identifier Type, select ROR.
Enter a valid ROR URL, e.g., https://ror.org/03vek6s52.
Save the dataset.
Navigate to the Metadata tab and click on the displayed ROR URL.
Observed Behavior
The ROR URL redirects to an invalid link:
https://ror.org/https://ror.org/03vek6s52.
This results in a 404 Page Not Found error due to duplication of the domain (ror.org).

Test with only the ROR Identifier (e.g., 03vek6s52):

Enter just the identifier (without the full URL).
Save the dataset and navigate to the Metadata tab.
The ROR Identifier is displayed as plain text and is not hyperlinked.
Expected Behavior
When a valid ROR URL is entered, the metadata tab should display and link to the correct ROR page, e.g., https://ror.org/03vek6s52.
When only the ROR Identifier is provided, the system should construct a valid URL (https://ror.org/{identifier}) and display it as a clickable hyperlink in the metadata tab.


https://github.com/user-attachments/assets/134b9306-c198-4d3d-9f8c-343bea78a54d

pdurbin commented:
> The ROR Identifier is displayed as plain text and is not hyperlinked.

Good catch. I plan to follow the pattern established in this PR to address this:

- #7979

jggautier commented:
Hi @pdurbin and @ofahimIQSS. I'm jumping in to make sure I understand how this should work, since I'm interviewing installation managers, starting this week, and planning usability testing of the changes we're making to this and another metadata field.

@ofahimIQSS, it's kind of fortuitous that you were able to share how you imagined that ROR identifiers would be handled. It's a bit of usability testing by someone with fresh eyes. Whereas if we had written about how the fields were designed and what to expect when the ROR IDs are displayed, you might not have been able to provide another example about what users will do and what they'll expect.

But I agree with Phil that the pattern established in https://github.com/IQSS/dataverse/pull/7979 - how the fields expect users to enter ORCIDs and they'll be displayed - should be followed here for RORs, too. To make things clearer for me, I'm including some mockups about that pattern.

@pdurbin, could you let me know if this is how it should work, based on what was done in https://github.com/IQSS/dataverse/pull/7979 for ORCIDs?

We expect users to select ORCID or ROR from the Identifier Type field's dropdown and enter the non-URL forms of ORCIDs (4 groups of alphanumeric characters separated by 3 dashes) and ROR IDs:

<kbd><img width="529" alt="Screenshot 2025-01-13 at 4 23 43 PM" src="https://github.com/user-attachments/assets/00736ae6-1605-49a3-bc0b-f2a30e4ced88" /></kbd>

Dataverse displays the ORCIDs and ROR IDs as links only when users do those two things for each of those identifiers:

<kbd><img width="389" alt="Screenshot 2025-01-13 at 4 25 51 PM" src="https://github.com/user-attachments/assets/87ef2e51-1e66-4555-a730-458de447ea1f" /></kbd>

On the other hand, when users select ORCID or ROR from the Identifier Type field's dropdown and enter the URL form of the ORCID or ROR, or anything else that isn't an ORCID (4 groups of alphanumeric characters separated by 3 dashes) or a ROR ID:

<kbd><img width="536" alt="Screenshot 2025-01-13 at 4 34 18 PM" src="https://github.com/user-attachments/assets/3593d6e8-5153-423c-b286-785a8b7e1f14" /></kbd>

 They're displayed as text and not as links:

<kbd><img width="389" alt="Screenshot 2025-01-13 at 4 34 23 PM" src="https://github.com/user-attachments/assets/a332edf2-4c61-4bb2-b979-2e417aaaeb3e" /></kbd>

pdurbin commented:
@jggautier right, the established pattern is that just the identifier (e.g. 03vek6s52) should be entered under Author Identifier rather than the full URL (e.g. https://ror.org/03vek6s52).

I added a note about this to the User Guide: b9d0146 . Here's a preview:

![Screenshot 2025-01-22 at 10 35 20 AM](https://github.com/user-attachments/assets/e1250501-8917-4150-a96a-cf5c6545fc26)

I've adjusted the code in this PR and have written tests to clarify this. I also marked the following issue to be closed when this PR is merged (the ROR link is working now, as shown in the screenshot below):

- #11118

When testing this PR, please also check the "Datacite" exporter because I had to touch that code. It should work the same as before, giving output like this...

`<nameIdentifier nameIdentifierScheme="ORCID" schemeURI="https://orcid.org">https://orcid.org/0000-0002-9528-9470</nameIdentifier>`

.. when you enter an ORCID properly, as discussed above and shown in the screenshot below.

In addition, you should now see output like this for RORs:

`<nameIdentifier nameIdentifierScheme="ROR" schemeURI="https://ror.org">https://ror.org/03vek6s52</nameIdentifier>`

I'm providing a screenshot below for how this looks in context.

![Screenshot 2025-01-22 at 10 04 49 AM](https://github.com/user-attachments/assets/c7bdc7cf-0796-4fda-9f15-90d6f5a0e600)
![Screenshot 2025-01-22 at 10 04 58 AM](https://github.com/user-attachments/assets/b12758a8-7fdf-4e42-bcaa-37b3df9aeb4c)
![Screenshot 2025-01-22 at 10 05 24 AM](https://github.com/user-attachments/assets/f1894b84-5454-4856-8f3f-b4bd9af90cfe)

qqmyers reviewed (approved):
Looks fine

ofahimIQSS commented:
Testing Passed - Merging PR

![image](https://github.com/user-attachments/assets/e2050bf9-319c-49c4-ba6e-154b5be36d89)

![image](https://github.com/user-attachments/assets/3718378b-0a3f-42a6-a202-fb64a0375162)
